nemo/collections/nlp/modules/common/megatron/transformer.py

```diff
@@ -632,6 +646,9 @@ def forward(
         if get_key_value:
             output = [output, presents]
 
+        if return_crossattention_scores or return_selfattention_scores:
+            output = [output, attention_probs]
```

Comment:
these simple condition tests are not robust. 
1. `attention_probs` is not declared when higher-level conditions are not applied. For example, self-attention scores are obtained when extra condition `self.layer_type != LayerType.retrieval_decoder_after_self_attn` should be met. 
2. the self-attention `attention_probs` could be overridden by cross-attention `attention_probs` (see line 604) in the above if-else logics.

So suggesting to declare `self_attention_probs` and `cross_attention_probs` separately, and initialize them to None at the beginning. Only return ones that are not None. That may lead to return both `self_attention_probs` and `cross_attention_probs`. Not sure if this is the use case we care about.

```suggestions
self_attention_probs = None
cross_attention_probs = None
...
if self.layer_type != LayerType.retrieval_decoder_after_self_attn:
   ...
   if return_selfattention_scores:
       attention_output, self_attention_probs = attention_output
   ...
if (
    self.layer_type == LayerType.decoder
    or self.layer_type == LayerType.retrieval_decoder
    or self.layer_type == LayerType.retrieval_encoder
    or self.layer_type == LayerType.retrieval_decoder_after_self_attn
):
    if (
        self.layer_type == LayerType.retrieval_decoder
        or self.layer_type == LayerType.retrieval_decoder_after_self_attn
    ):
       ...
    else:
        if return_crossattention_scores:
            attention_output, cross_attention_probs = attention_output
...
if self_attention_probs is not None:
   output = [output, self_attention_probs]
if cross_attention_probs is not None:
   output = [output, cross_attention_probs]
return output
```

Reply:
@blisc what do you think?

Reply:
Let's remove the option for `return_selfattention_scores`. The rest of the cross attention changes make sense

Reply:
ok, found a better solution,
```suggestions
        if return_crossattention_scores and return_selfattention_scores:
            raise NotImplementedError(
                "We can only return 1 of cross attention scores or self attention scores. Not both yet."
            )
        attention_probs = None
        ...
        if attention_probs is not None:
            output = [output, attention_probs]

        return output
```